core/src/main/java/org/wildfly/channel/ChannelSession.java

```diff
@@ -253,17 +277,34 @@ private ChannelImpl.ResolveLatestVersionResult findChannelWithLatestVersion(Stri
     }
 
     private Map<ChannelImpl, List<ArtifactCoordinate>> splitArtifactsPerChannel(List<ArtifactCoordinate> coordinates) {
+        long start = System.currentTimeMillis();
         Map<ChannelImpl, List<ArtifactCoordinate>> channelMap = new HashMap<>();
-        for (ArtifactCoordinate coord : coordinates) {
+
+        ForkJoinPool customThreadPool = new ForkJoinPool(versionResolutionParallelism);
+        ForkJoinTask<?> task = customThreadPool.submit(() -> coordinates.parallelStream().forEach(coord -> {
             ChannelImpl.ResolveLatestVersionResult result = findChannelWithLatestVersion(coord.getGroupId(), coord.getArtifactId(),
-                                                                                     coord.getExtension(), coord.getClassifier(), coord.getVersion());
+                    coord.getExtension(), coord.getClassifier(), coord.getVersion());
             ArtifactCoordinate query = new ArtifactCoordinate(coord.getGroupId(), coord.getArtifactId(), coord.getExtension(), coord.getClassifier(), result.version);
             ChannelImpl channel = result.channel;
-            if (!channelMap.containsKey(channel)) {
-                channelMap.put(channel, new ArrayList<>());
+            synchronized (ChannelSession.this) {
+                if (!channelMap.containsKey(channel)) {
+                    channelMap.put(channel, new ArrayList<>());
+                }
+                channelMap.get(channel).add(query);
             }
-            channelMap.get(channel).add(query);
+        }));
+
+        try {
+            task.get();
+        } catch (ExecutionException | InterruptedException e) {
+            throw new RuntimeException("Unable to resolve latest artifact versions.", e);
```

Comment:
`findChannelWithLatestVersion` throws `UnresolvedMavenArtifactException` if a stream cannot be matched, I think we should preserve and re-throw it

Reply:
@spyrkob thanks for checking on this. The `UnresolvedMavenArtifactException` extends the `RuntimeException`. I don't think I would be catching it, so that one should be normally propagated without any interference. Or do I miss something?

Reply:
IIRC, `task.get` wraps any exception that was thrown in the task in `ExecutionException`.
So we'd end up with a `RuntimeException` wrapping `ExecutionException`, wrapping `UnresolvedMavenArtifactException`